JSDemos/Demos/DataGrid/CustomNewRecordPosition/React/App.js

```diff
@@ -0,0 +1,116 @@
+import React from 'react';
+import DataGrid, {
+  Column, Editing, ValidationRule, Button, Toolbar, Item, Scrolling,
+} from 'devextreme-react/data-grid';
+import SelectBox from 'devextreme-react/select-box';
+import Guid from 'devextreme/core/guid';
+import { dataSource } from './data.js';
+
+const newRowPositionOptions = ['first', 'last', 'pageTop', 'pageBottom', 'viewportTop', 'viewportBottom'];
+const scrollingModeOptions = ['standard', 'virtual'];
+
+function App() {
+  const [newRowPosition, setNewRowPosition] = React.useState('viewportTop');
+  const [scrollingMode, setScrollingMode] = React.useState('standard');
+  const [changes, setChanges] = React.useState([]);
+  const [editRowKey, setEditRowKey] = React.useState(null);
+
+  const onOptionChanged = React.useCallback((e) => {
+    if (e.fullName === 'editing.changes') {
+      setChanges(e.value);
+    } else if (e.fullName === 'editing.editRowKey') {
+      setEditRowKey(e.value);
+    }
+  });
+
+  const onAddButtonClick = React.useCallback((e) => {
+    const key = new Guid().toString();
+    setChanges([{
+      key,
+      type: 'insert',
+      insertAfterKey: e.row.key,
+    }]);
+    setEditRowKey(key);
+  });
```

Comment:
Pass empty array as second parameter (deps) for useCallback to avoid updating on each render.